protomap/proto.go

```diff
@@ -361,21 +362,47 @@ func resolvedPath(basePath, annotatedPath *gpb.Path) *gpb.Path {
 	return np
 }
 
+// UnmapOpt marks that a particular option can be supplied as an argument
+// to the ProtoFromPaths function.
+type UnmapOpt interface {
+	isUnmapOpt()
+}
+
+// IgnoreExtraPaths indicates that unmapping should ignore any additional
+// paths that are found in the gNMI Notifications that dot not have corresponding
+// fields in the protobuf.
+//
+// This option is typically used in conjunction with path compression where there
+// are some leaves that do not have corresponding fields.
+func IgnoreExtraPaths() *ignoreExtraPaths { return &ignoreExtraPaths{} }
+
+type ignoreExtraPaths struct{}
+
+// isUnmapOpt marks ignoreExtraPaths as an unmap option.
+func (*ignoreExtraPaths) isUnmapOpt() {}
+
 // ProtoFromPaths takes an input ygot-generated protobuf and unmarshals the values that are specified in the map
 // vals into it, using prefix as the prefix to any paths within the vals map. The message, p, is modified in place.
 // The map, vals, must be keyed by the gNMI path to the field which is annotated in the ygot generated protobuf,
-// the complete path is taken to be prefix + the key found in the map.
-func ProtoFromPaths(p proto.Message, vals map[*gpb.Path]interface{}, prefix *gpb.Path) error {
+// the complete path is taken to be prefix + the key found in the map. The supplied opt control the unmarshalling behaviour.
+func ProtoFromPaths(p proto.Message, vals map[*gpb.Path]interface{}, protoPrefix *gpb.Path, opt ...UnmapOpt) error {
 	if p == nil {
 		return errors.New("nil protobuf supplied")
 	}
 
 	directCh := map[*gpb.Path]interface{}{}
 	for p, v := range vals {
-		// TODO(robjs): needs fixing for compressed schemas.
-		if len(p.GetElem()) == len(prefix.GetElem())+1 {
+		if len(p.GetElem()) == 1 {
 			directCh[p] = v
 		}
+		// TODO(robjs): it'd be good to have something here that tells us whether we are in
+		// a compressed schema. Potentially we should add something to the generated protobuf
+		// as a fileoption that would give us this indication.
+		if len(p.Elem) == 2 {
+			if p.Elem[len(p.Elem)-2].Name == "config" || p.Elem[len(p.Elem)-2].Name == "state" {
+				directCh[p] = v
+			}
+		}
```

Comment:
So prefix is no longer part of the path value in `vals`?

Reply:
I've restructured this so that there are two clear types as we discussed -- one for prefixes to the values, and one for where the protobuf supplied is not the root. I've made these two work together too. PTAL.